README.md
# VSC-GPU: A Selenium-based VS Code Extension For Local Edit and Cloud Compilation on WebGPU
![build](https://img.shields.io/badge/build-passing-green) ![license](https://img.shields.io/badge/license-gpl-green) ![release](https://img.shields.io/badge/release-1.5.0--beta-green) 
![engine](https://img.shields.io/badge/engine-nodejs-yellow) ![machine type](https://img.shields.io/badge/machine-win--mac--linux-yellow)

<img src='./images/readme_about/signature.png' width="100">

* [Your next WebGPU, is not WebGPU](#your-next-webgpu-is-not-webgpu)
   * [Introduction](#introduction)
   * [KEEP IT OPEN-SOURCE](#keep-it-open-source)
   * [Getting started](#getting-started)
      * [Create Your Working file](#create-your-working-file)
      * [Install ECE408 Remote Control Extension](#install-ece408-remote-control-extension)
      * [Install the shell dependencies](#install-the-shell-dependencies)
      * [Install the Chrome driver](#install-the-chrome-driver)
         * [Before started](#before-started)
         * [Check for the right version of Chrome](#check-for-the-right-version-of-chrome)
         * [Download the correct version of chromedriver](#download-the-correct-version-of-chromedriver)
      * [Install the other extension](#install-the-other-extension)
      * [Network connection](#network-connection)
   * [Working steps](#working-steps)
      * [Config](#config)
      * [Login](#login)
      * [Pull](#pull)
      * [Push](#push)
      * [Exit](#exit)
   * [Trouble-shooting](#trouble-shooting)
   * [Possible Improvements](#possible-improvements)
   * [Issue Report](#issue-report)
   * [Contributions](#contributions)
      * [Set up your working environment](#set-up-your-working-environment)
      * [Renew the Extension Version](#renew-the-extension-version)
   * [Acknowledgements](#acknowledgements)
   * [Release Notes](#release-notes)

## Introduction

Recently, programming on the Web is developing quickly, and [SaaS](https://en.wikipedia.org/wiki/Software_as_a_service) has never been so popular. In UIUC, [WebGPU](https://www.webgpu.net/) is developed for [ECE408 (Applied Parallel Programming)](https://ece.illinois.edu/academics/courses/ece408) students to write, compile, and debug their code. However, as the architecture for WebGPU is not yet fully developed, and because of the natural difficulty to render code on HTML pages, we can observe that many students complain about this interface and they try to use other methods to code.

On the other side, [VS Code](https://code.visualstudio.com/) has never been so popular because of its strong support for every language and its huge extension community. VS Code is also strong because it's **pure** editor, so it's compatible to all of the operating systems, including Windows, Mac, Linux, etc.

I hereby, introduce a VS Code Extension called [ECE 408 Remote Control](https://marketplace.visualstudio.com/items?itemName=JackBAI.ece408-remote-control), which allows students to use their code created locally, using VS Code as the editor with decent syntax highlighting systems, and use the extension for posting the code back onto the WebGPU website, automatically running the tests online, and show the results on the Chrome browser.

## KEEP IT OPEN-SOURCE


As a steady supporter of the [open source software movement](https://en.wikipedia.org/wiki/Free_software_movement), for most of the time I'd like to make my softwares open. Therefore, this extension is under the [The GNU General Public License](./LICENSE). If you want to make any form of edit to it, please make sure you keep it open source.

## Getting started

Before you can use the extension, there are some setting-up steps to follow.


### Create Your Working file

When starting, you should use the VS Code to open your `.cu` CUDA source , for example,

```shell
code /path/to/your/workdir/lab1.cu
```

Note that you **must** use `.cu` as the postfix of your working file, or the extension `vscode-cudacpp` will **not** be able to recognize the file as a valid CUDA C++ file.


### Install ECE408 Remote Control Extension

To set up the working environment, you need to install my extension first. After you install it, you can see the interface as shown below. Please click on the "security" and trust my extension button if it is not yet trusted.

![overall](./images/readme_about/overall.png)

### Install the shell dependencies

`If` you're currently using Mac, use 
```shell
brew install pbcopy
```

`If` you're currently using Windows, `| clip` should already be in your system PATH for `cmd`. You can open `cmd.exe` (not `powershell`) and type in `echo 1 | clip` to check.

### Install the Chrome driver

You also need to install `chromedriver` into your system `PATH`.

#### Before started

Before you install the chromedriver, you should have:
![before started](./images/readme_about/before_started.png)

Then you need to install the `chromedriver`. 

#### Check for the right version of Chrome

Please open the current Chrome (if you have) and check which version you're in. For example, mine is 94.0.4606.

![check for right version](./images/readme_about/check_for_right_version.png)

#### Download the correct version of chromedriver

Go to [the download website](https://chromedriver.chromium.org/) to download the correct version of ChromeDriver. If you're using Windows, make sure you download the Windows version. I'm using Mac, so I'll download the Mac version.

![correct version](./images/readme_about/correct_version.png)

After downloading, please unzip the file and move it to your system PATH.

![move into PATH](./images/readme_about/move_into_path.png)

Then please check whether you've successfully installed your chromedriver. If you see the feedback below, you're all set.

![examine](./images/readme_about/examine.png)

Notes:
-   You must download exactly the same version of `chromedriver` as your chrome browser if you want to use the `chromedriver` as your selenium driver. The chrome browser on your machine is likely to be updated automatically when a new version is released. In this case, please update to the newest version of `chromedriver` too.
-   I strongly recommend you to install the `chromedriver` into the system software directory immediately, i.e. it should appear in `/usr/local/bin/chromedriver`.

### Install the other extension

In your VS Code workspace, please install the following extensions:

-   `vscode-cudacpp`: CUDA C++ syntax highlight.

### Network connection

For students in China, make sure you’re equipped with a VPN so that you can gain a fast connection with the WebGPU server held in UIUC.

## Working steps

| Shortcut | Canonical Press   | Step | Frequency            |
| -------- | ----------------- | ---- | -------------------- |
| `f1`     | `> WebGPU: Login` | 1    | often 1 time         |
| `f2`     | `> WebGPU: Pull`  | 2    | often 1 time         |
| `f5`     | `> WebGPU: Push`  | 3    | often multiple times |
| `f12`    | `> WebGPU: Exit`  | 4    | often 1 time         |

Note that the **Canonical Press** means you need to firstly press the combined keys `shift+ctrl/cmd+p` on your keyboard.

### Config

After creating the working file (`lab2.cu` in my case), please use VS Code to open it. Make sure you’ve done all things mentioned in the *Getting Started* part. 

Then please open the `Settings` page in VS code. Detailedly, please go through `Code/Preferences/Settings/ECE408 Remote Control/`. Alternatively, you can just press `ctrl,` (win) or `command,` (mac).

For the `headless` config, please choose `false` for now, because the headless mode is still under test.

![config](images/readme_about/config.gif)

### Login

This part is fully automated. Just type in `f1` and everything is set. After this step you will be on the code editing page.

![login](images/readme_about/login.gif)

### Pull

This part is fully automated. Just type in `f2` and everything is set. After this step the code on the WebGPU page will be automatically added to your working file.

![pull](images/readme_about/pull.gif)

### Push

This part is fully automated. Just type in `f5` and everything is set. After this step your code in your working file will be automatically pushed onto the WebGPU website, saved, and also submitted for feedbacks.

![push](images/readme_about/push.png)

### Exit

This part is to simply exit the Chrome browser and shut down the Internet connection to keep your account safe.


## Trouble-shooting


| Error | Cause | Solution | 
| -------- | --------------- | ---- | 
| `config` does not exist     | Did not install `chromedriver`      | Install `chromedriver`    | 
| Installed `chromedriver` did not launch     | Chrome Version not right       |  Check the version of Chrome and download the corresponding `chromedriver`   | 
| Not able to connect to WebGPU | Internet failure | Use UIUC (or other legal) VPN if you're in China; go to a place where Internet connection is better |
| Automation stops | Asynchronization function failed | refresh your page by hand |
| `push` did not change content on WebGPU | Did not install `pbcopy` on your Mac, or wrong-typed config about 'mac' or 'win' | `brew install pbcopy` or press `f1` and re-config


## Possible Improvements

-   Try to change *selenium* to **headless** mode, and move the HTML rending part to local workspace.

## Issue Report

If you find an issue or have some more features to request, please reply on the [Github Issues Page](https://github.com/BiEchi/ece408-remote-control/issues). Please do NOT post it on CampusWire, which is dedicated for course-related contents.

## Contributions

If you want to contribute to the code, please contact Haob.19@intl.zju.edu.cn to send **pull requests** to the [GitHub Repo](https://github.com/BiEchi/ece408-remote-control).

Before you start, please make sure you’ve already completed all steps in the *requirements* part.

### Set up your working environment

If you want to test your code after you make refinements to my code, please **make sure** there are no bugs. To debug, use the framework provided by Microsoft VS Code.

```shell
npm install -g yo generator-code
yo code // then follow the default settings
```

### Renew the Extension Version

To renew versions on the VS Code Extension Marketplace, please type 

```shell
npm i vsce -g # if you did not install vsce
vsce package # if you want to make a package
vsce publish # if you want to publish immediately
```

## Acknowledgements

I need to express my greatest thanks to [Professor Kindratenko](mailto:kindrtnk@illinois.edu), who encouraged me to go on with this VS Code extension. I also need to thank [TA Andrew Schuh](mailto:aschuh@illinois.edu), who is also the maintainer of WebGPU, who provided me with the web API for the `pull` part of the extension. Gratitudes are also given to my peer student and teammate [Zhenzuo SI](mailto:zhenzuo.19@intl.zju.edu.cn), who encouraged me to develop such a tool.

Although all people mentioned above did not contributed directly to any part of this extension, they provided me with the greatest warmth and appreciation, without which I cannot develop such a functional tool. Thanks to all of you!


## Release Notes

For the Release Notes part, please refer to [CHANGELOG](./CHANGELOG.md) for details.


CHANGELOG.md
# Change Log

All notable changes to the "ece408-remote-control" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.



## [1.8.5] - 2022.1.30

### Changed

-   Changed icon.



## [1.8.2] - 2022.1.29

### Changed

-   Changed displayed name and icon.



## [1.8.2] - 2021.10.28

### Added

-   Added the system to avoid ssl failure interruption when executing the `pull` operation.



## [1.6.0] - 2021-10.24

### Removed

-   Becaise SSL is back to life on the server, we delete the corresponding solver code.

### Changed

-   Due to ambiguity, we change the name to **VSC-WebGPU** from VSC-GPU.



## [1.5.2] - 2021-10.14

### Changed

-   The name from `ECE408 Remote Control` to `VSC-GPU`.



## [1.5.0] - 2021-10.14

### Added

-   Detailed process feedback on the OUTPUT channel.
-   Synchronization on `promise` objects.
-   The option to choose to be `headless` or not. (not stable).
-   The browser open/close automation.

### Changed

-   Resolved potential issues due to asynchronization.



## [1.2.0] - 2021-10-14

### Added

-   A more strict method for detecting whether already accessed the proposed website.

### Changed

-   The `vscode.window.showInformation()` method to `startOutputChannel()` method.



## [1.1.0] - 2021-10-13

### Added

-   The `config` file to avoid typing in information every time.

### Changed

-   The README file about the `config` part.
-   The route for MP 5.1 and MP 5.2.
-   The shortcuts for `login` and `pull` commands due to delete of `config` command.

### Removed

-   The `config` command.



## [1.0.4] - 2021-10-10

### Added

-   Trouble-shotting part in [README](./README.md).
-   Production support for Mac users.



## [1.0.1] - 2021-10-10

### Added

-   Production support for Windows users.



## [1.0.0] - 2021-10-10 {MILESTONE}

### Added

-   The automated pushing system.
-   Version Navigation.
-   Complete installation GIF files.
-   CHANGELOG.
-   Added `.then()` functionalities to improve robustness.
-   Added `push` functionality.
-   The acknowledgement part in `README.md`.
-   **The CP-CI integration onto GitHub.**

###  Changed

-   The extension icon.
-   Fixed bugs in the `config` part.
-   Merged unnecessary functions in `push`.

### Removed

-   The HTML downloading and  rendering part in `push` (will be re-added later).
-   The code reading part in `push`.



## [0.9.6] - 2021-10-7

### Changed

-   The extension icon.



## [0.9.5] - 2021-10-6

### Added

-   The code reading part in `push`.
-   The HTML downloading and rendering part in `push`.

### Changed

-   Fixed some bugs in `pull`.
-   Revised the workflow and almost completed `README.md`.



## [0.9.2] - 2021.10.1

### Changed

-   Fixed the bugs on the beginning of the code.



## [0.9.0] - 2021.9.27

### Added

-   Completed `config` execution, which allows users to input their account and password.



## [0.8.0] - 2021.9.24

### Added

-   Completed `pull` execution.

### Changed

-   Reconstructed the functions using subroutines and system decomposition.



## [0.7.0] - 2021.9.21 {MILESTONE}

### Added

-   Completed `login` execution.
-   Completed `exit` execution.
-   `pull` sledding comes to 50/100 %.
-   `push` sledding comes to 40/100 %.



## [0.4.0] - 2021.9.16

### Added

-   Initialized `push` execution.

### Changed

-   Updated the icon for the extension.



## [0.3.0] - 2021.9.14

### Added

-   Added the GitHub Repository for the extension source file VCS.



## [0.2.0] - 2021.9.13

### Changed

-   Completed the login process.
-   Next step is to write the files using DOM in VS Code.



## [0.1.0] - 2021.9.10

### Added

-   Initial ‘release’ of the project.
-   Created the marketplace position.



extension.js
//=============================================================================
//
//   VS Code Extension for the lecture
//
//   "Applied Parallel Programming"
//
//   Jack BAI
//
//   Copyright (C) timeout_time21 by Hepta Workshop
//
//-----------------------------------------------------------------------------
//
//                                License
//
//   This program is free software; you can redistribute it and/or
//   modify it under the terms of the GNU General Public License
//   as published by the Free Software Foundation; either version 2
//   of the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with this program; if not, write to the Free Software
//   Foundation, Inc., 51 Franklin Street, Fifth Floor,
//   Boston, MA  02110-1301, USA.
//
//=============================================================================

const vscode = require('vscode');


/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	//=============================================================================
	//
	//  GLOBAL PARAMETERS
	//
	//=============================================================================

	// This part of code will only be executed once when your extension is activated
	const timeout_time = 7;
	var account = vscode.workspace.getConfiguration('ece408').get('account');
	var passwd = vscode.workspace.getConfiguration('ece408').get('password');
	var num_lab = vscode.workspace.getConfiguration('ece408').get('lab_num');
	var machine = vscode.workspace.getConfiguration('ece408').get('machine');
	var headless_flag = vscode.workspace.getConfiguration('ece408').get('headless');
	
	var addr_list = ['9999', '10001', '10002', '10003', '10010', '10004', '10005', '10011', '10124'];
	var addr;
	switch (num_lab)
	{
		case '0':
			addr = addr_list[0];
			break;
		case '1':
			addr = addr_list[1];
			break;
		case '2':
			addr = addr_list[2];
			break;
		case '3':
			addr = addr_list[3];
			break;
		case '4':
			addr = addr_list[4];
			break;
		case '5.1':
			addr = addr_list[5];
			break;
		case '5.2':
			addr = addr_list[6];
			break;
		case '6':
			addr = addr_list[7];
			break;
		case '7':
			addr = addr_list[8];
			break;
		default:
			break;
	}

	var MP_xpath_list = ['4', '5', '6', '7', '8', '9', '10', '11', '12'];
	var MP_xpath;
	switch (num_lab)
	{
		case '0':
			MP_xpath = MP_xpath_list[0];
			break;
		case '1':
			MP_xpath = MP_xpath_list[1];
			break;
		case '2':
			MP_xpath = MP_xpath_list[2];
			break;
		case '3':
			MP_xpath = MP_xpath_list[3];
			break;
		case '4':
			MP_xpath = MP_xpath_list[4];
			break;
		case '5.1':
			MP_xpath = MP_xpath_list[5];
			break;
		case '5.2':
			MP_xpath = MP_xpath_list[6];
			break;
		case '6':
			MP_xpath = MP_xpath_list[7];
			break;
		case '7':
			MP_xpath = MP_xpath_list[8];
			break;
		default:
			break;
	}
	
	// flag to determine whether this is the first time for logging in
	var first_time_login = true;
	var output_channel = vscode.window.createOutputChannel("WebGPU");
	output_channel.show();
	output_channel.appendLine("If any sub-process (with [num%] on the left) exceeds 10 seconds, please launch the command again.");
	output_channel.appendLine("Because the headless mode is still under testing, please don't flag the 'headless' config for now.");

	// Initialization
	const {By, Key, until} = require('selenium-webdriver');
	var webdriver = require('selenium-webdriver'),
		chrome = require('selenium-webdriver/chrome'),
		options = new chrome.Options();
	if (headless_flag == true)
		options.addArguments('--headless');

	var driver = new webdriver.Builder()
		.forBrowser('chrome')
		.withCapabilities(options)
		.build();


	driver.manage().window().minimize();

	//=============================================================================
	//
	//  HELPER FUNCTIONS
	//
	//=============================================================================

	function solve_potential_bad_ssl() {
		// website error handling (temporary)
		var advanced_button = driver.wait(until.elementLocated(By.xpath('//*[@id="details-button"]'), timeout_time));
		advanced_button.click()
		.then(function(){
			output_channel.appendLine("[LOGIN/29%] SSL Certification failure detected. Trying to solve it...");
		});
	
		var proceed_button = driver.wait(until.elementLocated(By.xpath('//*[@id="proceed-link"]'), timeout_time));
		proceed_button.click()
		.then(function(){
			output_channel.appendLine("[LOGIN/55%] SSL failure handled. Proceeding to the normal login page.");
		})
	}

	function enter_information() {
		var login_button = driver.wait(until.elementLocated(By.xpath('//*[@id="content"]/div/div/div/div[3]/a[2]/div'), timeout_time));
		login_button.click();

		var account_box = driver.wait(until.elementLocated(By.xpath('//*[@id="user_name"]'), timeout_time));
		account_box.sendKeys(account);

		var passwd_box = driver.wait(until.elementLocated(By.xpath('//*[@id="password"]'), timeout_time));
		passwd_box.sendKeys(passwd)
		.then(function(){
			output_channel.appendLine("[LOGIN/74%] Finished entering identification information and redirecting to the lab page.");
		});
	}

	function go_to_lab_page() {
		driver.get('https://www.webgpu.net')
		.then(function(){
			output_channel.appendLine("[LOGIN/100%] Successfully logged in and have accessed Lab" + num_lab.toString() + ".");
			// we use the approach to click on the tabs to load the code
			var more_tab = driver.wait(until.elementLocated(By.xpath('/html/body/nav/div/div/button'), timeout_time));
			more_tab.click()
			var dropdown_toggle = driver.wait(until.elementLocated(By.xpath('/html/body/nav/div/nav/ul[1]/li[1]/a'), timeout_time));
			dropdown_toggle.click();
			var mp_button = driver.wait(until.elementLocated(By.xpath('/html/body/nav/div/nav/ul[1]/li[1]/ul/li[' + MP_xpath + ']/a'), timeout_time));
			mp_button.click();
			var code_tab = driver.wait(until.elementLocated(By.xpath('//*[@id="code-tab"]'), timeout_time));
			code_tab.click()
		})
	}

	function save_file(code){
		const fs = require('fs');
		const currentlyOpenTabfilePath = vscode.window.activeTextEditor.document.fileName;
		fs.truncateSync(currentlyOpenTabfilePath);
		fs.appendFileSync(currentlyOpenTabfilePath, code);
	}

	function copy_to_clipboard(){
		var exec = require('child_process').exec;
		var spawn = require('child_process').spawn;
		const currentlyOpenTabfilePath = vscode.window.activeTextEditor.document.fileName;
		if (machine == 'mac'){
			exec(`cat ${currentlyOpenTabfilePath} | pbcopy`);
		} else {
			spawn('cmd.exe', ['/c', `type ${currentlyOpenTabfilePath} | clip`]);
		}
	}

	function redirect_stdout(){
		output_channel.appendLine("### Timer Output ###");
		driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[4]/div[2]'), timeout_time))
		.then(function(timer_output_block){
			return timer_output_block.getText();
		}).then(function(timer_output_text){
			output_channel.appendLine(timer_output_text);

			output_channel.appendLine("### Logger Output ###");
			driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[5]/div[2]'), timeout_time))
			.then(function(logger_output_block){
				return logger_output_block.getText();
			}).then(function(logger_output_text){
				output_channel.appendLine(logger_output_text);
				go_to_lab_page();
			})
		});
	}

	function redirect_stderr(){
		output_channel.appendLine("### Error Message ###");
		driver.wait(until.elementLocated(By.xpath('/html/body/div[4]/h2'), timeout_time))
		.then(function(error_name_block){
			return error_name_block.getText();
		}).then(function(error_name_text){
			output_channel.appendLine(error_name_text.toUpperCase());

			driver.wait(until.elementLocated(By.xpath('/html/body/div[4]/p'), timeout_time))
			.then(function(error_content_block){
				return error_content_block.getText();
			}).then(function(error_content_text){
				output_channel.appendLine(error_content_text);
				var cancel_button = driver.wait(until.elementLocated(By.xpath('/html/body/div[4]/div[7]/button'), timeout_time));
				cancel_button.click();
			})
		})
	}

	function feedback(){
		output_channel.appendLine("[PUSH/41%] Code pasted to WebGPU. Running your code...");

		var stdout_symbol = driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[3]/div[1]/div/h3'), timeout_time));
		var stderr_symbol = driver.wait(until.elementLocated(By.xpath('/html/body/div[4]/h2'), timeout_time));
		
		stdout_symbol.click()
		.then(function(){
			output_channel.appendLine("[PUSH/79%] Compilation successful! Redirecting to LOGIN...");
			redirect_stdout();
		});

		stderr_symbol.click()
		.then(function(){
			output_channel.appendLine("[PUSH/79%] Compilation failed. Redirecting to LOGIN...");
			redirect_stderr();
		});

		driver.manage().window().minimize();
	}


	//=============================================================================
	//
	//  CALLER FUNCTIONS
	//
	//=============================================================================

	let login_process = vscode.commands.registerCommand('ece408-remote-control.login', function () {
		if (num_lab == 'null') {
			output_channel.appendLine("Error: please input your account and password first!");
			return;
		}

		// Certify the website
		output_channel.appendLine("[LOGIN/7%] Logging into your WebGPU account and accessing Lab" + num_lab.toString() + "...");
		driver.get('https://www.webgpu.net/');
		solve_potential_bad_ssl();
		if (first_time_login == true)
		{
			enter_information();
			var confirm_login_button = driver.wait(until.elementLocated(By.xpath('/html/body/div/div/div/form/div[3]/div/button'), timeout_time));
			confirm_login_button.click()
			.then(function(){
				output_channel.appendLine("[LOGIN/88%] Successfully pushed the indentification information to the server.");
				go_to_lab_page();
				first_time_login = false;
			})
		} else {
			driver.wait(function() {
				return driver.findElement(By.xpath('//*[@id="content"]/div/div')).isDisplayed();
			}, timeout_time)
			.then(function() {
				output_channel.appendLine("[LOGIN/88%] Successfully pushed the indentification information to the server.");
				go_to_lab_page();
			})
		}
	});

	let pull_process = vscode.commands.registerCommand('ece408-remote-control.pull', function () {
		// get the code
		output_channel.appendLine("[PULL/22%] Pulling the latest code on WebGPU from Lab " + num_lab.toString() + ".");
		var code_editor;
		var code = "";
		driver.get('https://www.webgpu.net/mp/' + addr + '/program/')
		.then(function(){
			// potential SSL error here
			return driver.getCurrentUrl();
		}).then(function(currentUrl){
			if (currentUrl.toString() == 'https://www.webgpu.net/mp/' + addr + '/program/') {
				solve_potential_bad_ssl();
				enter_information();
				var confirm_login_button = driver.wait(until.elementLocated(By.xpath('/html/body/div/div/div/form/div[3]/div/button'), timeout_time));
				confirm_login_button.click()
				.then(function(){
					driver.get('https://www.webgpu.net/mp/' + addr + '/program/')
					.then(function(){
						output_channel.appendLine("[PULL/59%] Redirecting to the source code page.");
						code_editor = driver.wait(until.elementLocated(By.xpath('/html/body/pre'), timeout_time))
						code = code_editor.getText();
						code.then(function(code) {
							// save the raw data and overwrite the lab project
							save_file(code);
							output_channel.appendLine("[PULL/88%] Successfully saved the file. Redirecting to LOGIN...");
							// go through the login subroutine again
							go_to_lab_page();
						})
					})
				})
			} else {
				output_channel.appendLine("[PULL/59%] Redirecting to the source code page.");
				code_editor = driver.wait(until.elementLocated(By.xpath('/html/body/pre'), timeout_time));
				code = code_editor.getText();
				code.then(function(code) {
					// save the raw data and overwrite the lab project
					save_file(code);
					output_channel.appendLine("[PULL/88%] Successfully saved the file. Redirecting to LOGIN...");
					// go through the login subroutine again
					go_to_lab_page();
				})
			}
		})
	});

	let push_process = vscode.commands.registerCommand('ece408-remote-control.push', function () {
		output_channel.appendLine("[PUSH/14%] Reminder: please save your .cu file before you run.");
		// copy the content in the workspace to the clipboard
		copy_to_clipboard();

		// Move the cursor to the first line of code and click 
		var code_line = driver.wait(until.elementLocated(By.xpath('//*[@id="code"]/div[1]/div[2]/div/span/div/div[6]/div[1]/div/div/div/div[5]/div[1]/pre/span/span'), timeout_time));
		var compile_button = driver.wait(until.elementLocated(By.xpath('//*[@id="code"]/div[1]/div[1]/div/div[2]/div[1]/div'), timeout_time));
		var all_button = driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div/div/div[2]/div[1]/div[1]/div/div[2]/div[1]/ul/li[8]/a'), timeout_time));
		const actions = driver.actions();


		driver.manage().window().maximize();

		if (machine == 'mac'){
			actions
			.click(code_line)
			.keyDown(Key.COMMAND)
			.sendKeys('a')
			.sendKeys('v')
			.sendKeys('s')
			.keyUp(Key.COMMAND)
			.click(compile_button)
			.click(all_button)
			.perform()
			.then(function(){
				// download feedback
				feedback();
			});
		} else {
			actions
			.click(code_line)
			.keyDown(Key.CONTROL)
			.sendKeys('a')
			.sendKeys('v')
			.sendKeys('s')
			.keyUp(Key.CONTROL)
			.click(compile_button)
			.click(all_button)
			.perform()
			.then(function(){
				// download feedback
				feedback();
			})
		}
	});

	let exit_process = vscode.commands.registerCommand('ece408-remote-control.exit', function () {
		driver.quit();
		output_channel.appendLine("You've successfully exited your account!");
	});

	for (let item in {login_process, pull_process, push_process, exit_process})
		context.subscriptions.push(item);
}

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}
